### PR TITLE
Compatibility with Rails 5.2

### PR DESCRIPTION
**Pull request by bogdan**
This PR adapts the new AR migrations source code and makes the gem work the same way it worked on previous versions of rails.

I didn't make use of the new multi-database migrations functionality. It requires a lot of strategic decisions on how migrations and structure should be stored.
I think this is good 1st iteration and it makes it possible to upgrade to rails 5.2 painlessly.

### CODE REVIEW

File: lib/active_record_shards/default_slave_patches.rb [hunk 1]
```diff
@@ -100,7 +100,10 @@ def on_slave_unless_tx
 
     module ActiveRelationPatches
       def self.included(base)
-        [:calculate, :exists?, :pluck, :find_with_associations].each do |m|
+        [
+          :calculate, :exists?, :pluck,
+          ActiveRecord::VERSION::STRING >= '5.2.0' ? :load : :find_with_associations
```
Comment (bquorning) [inline lib/active_record_shards/default_slave_patches.rb]: The commit message for rails/rails@412db710dfa6ed84654068576b1841966d7f89b2 is “Use `apply_join_dependency` instead of meaningless named `find_with_associations`” – I wonder why `:load` was chosen here instead of `:apply_join_dependency`.

File: lib/active_record_shards/migration.rb [hunk 2]
```diff
@@ -1,25 +1,31 @@
...
+          send("#{method}_without_sharding", *args)
+        end
+        ActiveRecord::Base.on_all_shards do
+          send("#{method}_without_sharding", *args)
         end
-        alias_method :"#{m}_without_sharding", m.to_sym
-        alias_method m.to_sym, :"#{m}_with_sharding"
       end
+      alias_method :"#{method}_without_sharding", method.to_sym
+      alias_method method.to_sym, :"#{method}_with_sharding"
```
Comment (bquorning) [inline lib/active_record_shards/migration.rb]: This patch (lines 4–16) could be extracted into a separate commit – or a separate pull request, even.

I am thinking, perhaps it would be better not to use metaprogramming when we are only overriding 2 methods. E.g.:

```rb
    def run_with_sharding(*args)
      ActiveRecord::Base.on_shard(nil) { run_without_sharding(*args) }
      ActiveRecord::Base.on_all_shards { run_without_sharding(*args) }
    end
    alias_method :run_without_sharding, :run
    alias_method :run, :run_with_sharding

    def migrate_with_sharding(direction)
      ActiveRecord::Base.on_shard(nil) { migrate_without_sharding(direction) }
      ActiveRecord::Base.on_all_shards { migrate_without_sharding(direction) }
    end
    alias_method :migrate_without_sharding, :migrate
    alias_method :migrate, :migrate_with_sharding
```

File: test/helper.rb [hunk 3]
```diff
@@ -89,6 +89,20 @@ def table_has_column?(table, column)
...
+    if ActiveRecord::VERSION::MAJOR >= 4
+      migrations = if ActiveRecord::VERSION::STRING >= "5.2.0"
+                     ActiveRecord::MigrationContext.new(migration_path).migrations
+                   else
+                     ActiveRecord::Migrator.migrations(migration_path)
+                   end
+      ActiveRecord::Migrator.new(direction, migrations, target_version)
+    else
+      ActiveRecord::Migrator.new(direction, migration_path, target_version)
+    end
```
Comment (bquorning) [inline test/helper.rb]: Would it be too much duplication with e.g.
```rb
    if ActiveRecord::VERSION::STRING >= "5.2.0"
      migrations = ActiveRecord::MigrationContext.new(migration_path).migrations
      ActiveRecord::Migrator.new(direction, migrations, target_version)
    elsif ActiveRecord::VERSION::MAJOR >= 4
      migrations = ActiveRecord::Migrator.migrations(migration_path)
      ActiveRecord::Migrator.new(direction, migrations, target_version)
    else
      ActiveRecord::Migrator.new(direction, migration_path, target_version)
    end
```

File: gemfiles/rails5.2.gemfile [hunk 4]
```diff
@@ -1,4 +1,3 @@
 eval_gemfile 'common.rb'
 
-gem 'activerecord', '5.2.0.beta2'
-gem 'mysql2', '~> 0.4.4'
+gem 'activerecord', '~> 5.2.0'
```
Comment (bquorning) [inline gemfiles/rails5.2.gemfile]: Nitpick: Missing newline.

File: lib/active_record_shards/default_slave_patches.rb (hunk 1 above)
Comment (bogdan) [reply lib/active_record_shards/default_slave_patches.rb]: `apply_join_dependency` is an internal method. We have to do the change because internal methods tend to be changed with no backward compatibility. Also `apply_join_dependency` is used in many places - when we hack it, we are not sure which public methods are actually get hacked.

File: test/helper.rb (hunk 3 above)
Comment (bogdan) [reply test/helper.rb]: Rubocop suggested me this style. Build fails with your code.

File: lib/active_record_shards/default_slave_patches.rb (hunk 1 above)
Comment (bquorning) [reply lib/active_record_shards/default_slave_patches.rb]: That makes sense. What about `:load`, that sounds like a generic method too?

Does `:load` exist in older versions of Rails, so we could just patch that method instead of `find_with_associations`?

File: test/helper.rb (hunk 3 above)
Comment (bquorning) [reply test/helper.rb]: Really? I guess I should go fix RuboCop then :-D

Comment (bogdan) [reply test/helper.rb]: I can use your code snippet if you would suggest me the way to ignore rubocop rule for this code.

Comment (bquorning) [reply test/helper.rb]: I don’t see RuboCop complaining on my code. What error are you seeing, exactly?

**Comment by bogdan** [comment]
@bquorning all done.

**Comment by bquorning** [comment]
I am still wondering about that `:load` vs `:find_with_associations` change. Could we remove the conditional and always just patch `:load`?

Was it an oversight in #64 to not patch `:find_with_associations` instead of `:load`, or does that have extra consequences that aren’t tested. What do you think, @grosser ?

**Comment by bogdan** [comment]
I removed it and the tests failed on rails 3.2. Now, I've changed the condition to have exceptional case for version 3.2 instead of 5.2

**Comment by bquorning** [comment]
Thanks @bogdan 

> does that have extra consequences that aren’t tested

Waiting for @grosser (or someone else) to chime in on this.

**Comment by pschambacher** [comment]
Go for it. Let’s merge!

File: test/migrator_test.rb [hunk 5]
```diff
@@ -29,48 +28,48 @@
...
-    migration_path = File.join(File.dirname(__FILE__), "/failure_migration")
-
-    assert failure_migration_pending?(migration_path)
+    assert failure_migration_pending?('failure_migration')
     begin
-      ActiveRecord::Migrator.migrate(migration_path)
-    rescue
+      migrator(:up, 'failure_migration').migrate
+    rescue => e
+      unless e.message.include?("FAIL")
```
Comment (grosser) [inline test/migrator_test.rb]: might be nice to make this more specific and add a comment on what exactly it's catching

File: lib/active_record_shards/migration.rb [hunk 6]
```diff
@@ -1,25 +1,32 @@
...
+    end
+    alias_method :run_without_sharding, :run
+    alias_method :run, :run_with_sharding
+
+    def migrate_with_sharding
+      ActiveRecord::Base.on_shard(nil) { migrate_without_sharding }
+      ActiveRecord::Base.on_all_shards { migrate_without_sharding }
+    end
+    alias_method :migrate_without_sharding, :migrate
+    alias_method :migrate, :migrate_with_sharding
```
Comment (grosser) [inline lib/active_record_shards/migration.rb]: prefer using a prepended module

File: lib/active_record_shards/migration.rb [hunk 7]
```diff
@@ -1,25 +1,32 @@
...
-          ActiveRecord::Base.on_shard(nil) do
-            send("#{m}_without_sharding", *args)
-          end
-          ActiveRecord::Base.on_all_shards do
-            send("#{m}_without_sharding", *args)
-          end
-        end
-        alias_method :"#{m}_without_sharding", m.to_sym
-        alias_method m.to_sym, :"#{m}_with_sharding"
+    def self.migration_context
```
Comment (grosser) [inline lib/active_record_shards/migration.rb]: if we made that method up, then prefer prefixing `shards_` or so to make that obvious

File: lib/active_record_shards/migration.rb (hunk 6 above)
Comment (bquorning) [reply lib/active_record_shards/migration.rb]: prepend used to conflict with New Relic’s gem. Dunno if that is still the case.

Anyhow, I think changing our aliasing strategy is out of scope of this PR.

**Comment by bquorning** [comment]
> Let’s merge!

Please give me some more confidence in that `load` method patch :-)

File: lib/active_record_shards/migration.rb (hunk 6 above)
Comment (pschambacher) [reply lib/active_record_shards/migration.rb]: Yes let's not change this to prepend as part of this PR.

Comment (bogdan) [reply lib/active_record_shards/migration.rb]: This gem is using the old ruby 1.x style for everything, so I would fix it at scale somewhere else...

**Comment by bogdan** [comment]
> Please give me some more confidence in that load method patch :-)

@bquorning I am not using this feature myself, but if you do, you can:

* use this branch in your app
* run your test suite
* analyze the test.log to see if all requests for certain models are going to slave

**Comment by bquorning** [comment]
👍 I’ll merge this in now and make a beta release real soon, so we can test this code against our applications.

**Comment by bquorning** [comment]
v3.12.0.beta1 is ready.
